libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp

```diff
@@ -197,6 +197,18 @@ void NavEKF2_core::readMagData()
             }
         }
 
+        // detect changes to magnetometer offset parameters and reset states
+        Vector3f nowMagOffsets = _ahrs->get_compass()->get_offsets(magSelectIndex);
+        bool changeDetected = lastMagOffsetsValid && (!is_equal(nowMagOffsets.x,lastMagOffsets.x) || !is_equal(nowMagOffsets.y,lastMagOffsets.y) || !is_equal(nowMagOffsets.z,lastMagOffsets.z));
+        if (changeDetected) {
+            // zero the learned magnetometer bias states
+            stateStruct.body_magfield.zero();
+            // clear the measurement buffer
+            storedMag.reset();
+        }
+        lastMagOffsets = nowMagOffsets;
+        lastMagOffsetsValid = true;
+
```

Comment:
Instead of doing this every time we get the mag data, couldn't we do the other way? In the (rare) cases the offsets are set we let EKF reset the states.